lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,344 @@
+      }
+      instances.push_back(otherInstance);
+    }
+
+    return instModule;
+  }
+
+  /// If the `argNames` attribute exists, return it.
+  SmallVector<StringAttr> getInstanceArgNames(HWInstanceLike inst) {
+    SmallVector<StringAttr> argNames;
+    if (auto args = inst->getAttrOfType<ArrayAttr>("argNames")) {
+      auto attr = args.getAsRange<StringAttr>();
+      argNames.insert(argNames.begin(), attr.begin(), attr.end());
+    }
+    return argNames;
+  }
+
+  /// Combine two consecutive clock_gate ops, into one.
+  void collapseConsecutiveClockGates(ClockGateOp driverGate,
+                                     ClockGateOp userGate) {
+    mlir::OpBuilder builder(context);
+    builder.setInsertionPoint(userGate);
+    Value enable = getEnable(driverGate);
+    enable = builder.create<comb::AndOp>(userGate->getLoc(), enable,
+                                         userGate.getEnable());
+    userGate.getInputMutable().assign(driverGate.getInput());
+    userGate.getEnableMutable().assign(enable);
+  }
+
+  /// If the clock_gate op has `test_enable`, combine it with the `enable` and
+  /// return the final enable.
+  Value getEnable(ClockGateOp clkGate) {
+    Value enable = clkGate.getEnable();
+    // If there is testEnable, or it.
+    if (auto testEnable = clkGate.getTestEnable()) {
+      OpBuilder builder(clkGate.getContext());
+      builder.setInsertionPointAfter(clkGate);
+      enable =
+          builder.create<comb::OrOp>(clkGate->getLoc(), enable, testEnable);
+    }
+    return enable;
+  }
+
+  void runOnOperation() override;
+
+private:
+  using ClockEnablePortPairs = DenseMap<unsigned, unsigned>;
+  InstanceGraph *graph;
+  DenseSet<Operation *> opsToErase;
+  mlir::MLIRContext *context;
+  DenseMap<HWModuleOp, ClockEnablePortPairs> moduleClockGateEnableMap;
+};
+
+void SinkClockGatesPass::runOnOperation() {
+  graph = &getAnalysis<circt::hw::InstanceGraph>();
+  // A setvector is required for a deterministic output.
+  SetVector<HWInstanceLike> instancesWithGatedClk;
+  DenseSet<Operation *> opsWithGatedClock;
+  context = getOperation().getContext();
+
+  // Record all the instances that have clock_gate as an input. This will be
+  // used to push the clock_gate down through the hierarchy to its users.
+  // This also collapses consecutive clock gates.
+  auto findInstancesWithGatedClock = [&](HWModuleOp hwModule) {
+    hwModule->walk([&](ClockGateOp clkGate) {
+      for (auto &use : clkGate->getUses()) {
+        auto *useOp = use.getOwner();
+        if (auto inst = dyn_cast<HWInstanceLike>(useOp))
+          instancesWithGatedClk.insert(inst);
+        else if (auto userGate = dyn_cast<ClockGateOp>(useOp)) {
+          collapseConsecutiveClockGates(clkGate, userGate);
+        } else {
+          opsWithGatedClock.insert(useOp);
+        }
+      }
+    });
+  };
+  // Seed phase, find the instances to start the traversal from.
+  // These instances have a gated clock as input.
+  getOperation()->walk(
```

Comment:
Can this use an instance graph or something to find the modules, instead of `walk` through the entire design looking for HWModuleOp's?

Reply:
Right, thanks for catching this, moved to iterate over instance graph nodes.